### DISCUSSION TITLE
When does HttpClient with HTTP/1.x open new TCP connections?

### DISCUSSION THREAD

**Discussion by alexandrehtrb** [dotnet/runtime#99532] [Q&A] [answered]
Hello,

Consider a single HttpClient being used to make 100 concurrent HTTP/1.1 calls to a server. HTTP/1.1 uses persistent TCP connections and by using a singleton HttpClient, we're reusing the connection pool. HTTP/1.1 has head-of-line blocking (a request can't be sent unless the previous has finished, on a single TCP connection).

Will always only one TCP connection be used for each server, in a singleton HttpClient? Or does it open new parallel TCP connections if there are too many calls to a server?

**Comment by stephentoub** [comment]
By default if there's no connection available in the pool, it'll create a new one, so in your scenario, if there are 100 concurrent requests that take a meaningful amount of time, you'll end up with ~100 connections. If you set MaxConnectionsPerServer on the handler, and that many connections already exists to the server with that handler, then instead of creating a new connection it'll wait for an existing one to become available.

**Reply by alexandrehtrb** [reply]
But if there is already a "warmed-up" connection, then it will only one connection in the pool?

**Reply by stephentoub** [reply, marked as answer]
If a request was made and completed, the connection it was using will be returned to the pool. If another request comes along and finds an available connection in the pool, it'll use it.

**Reply by alexandrehtrb** [reply]
> If another request comes along and finds an <ins>available</ins> connection in the pool, it'll use it.

- 1 warmed-up connection exists in the pool. This connection is idle (not being used).
- 5 concurrent calls to the server arrive at the same time for the HttpClient to send them.
- 1 call will reuse the existing connection, the 4 others will spawn 4 new TCP connections, because there are no other pre-existing available connections.

Is that it?

**Reply by MihaZupan** [reply]
Yes

**Comment by hyke123** [comment]
Hi @stephentoub  We're running into an issue when using a single HttpClient to upload files to S3. If multiple requests are triggered simultaneously, sometimes we get the error: "System.Net.Http.HttpIOException: The response ended prematurely. (ResponseEnded) at System.Net.Http.HttpConnection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)." I suspect it might be because the S3 server is closing the reused connections. Do you think the "SocketsHttpHandler" would solve this, or do you have any other suggestions?

**Reply by huoyaoyuan** [reply]
> I suspect it might be because the S3 server is closing the reused connections.

No, it should be unrelated. HTTP/1.1 doesn't support duplex and simultaneously requests should be on different TCP connections, not reused ones.

> Do you think the "SocketsHttpHandler" would solve this, or do you have any other suggestions?

`SocketsHttpHandler` is the default underlying implementation of `HttpClient`.

You can try to collect packets, or read the documentation about any limitations. It's unrelated to this thread. If you discovered further information, you can create new discussion of issue.

**Reply by hyke123** [reply]
Ok, Thanks for your reply!
